cfgov/unprocessed/js/modules/TreeDomModel.js

```diff
@@ -0,0 +1,86 @@
+import Tree from './Tree';
+
+/**
+ * TreeDomModel
+ * @class
+ *
+ * @classdesc A tree data structure.
+ * Trees have one root node, and child nodes that branch.
+ * Like:
+ * @returns {TreeDomModel} An instance.
+ */
+function TreeDomModel() {
+
+  let _tree;
+
+  /**
+   * @param {HTMLNode} dom - A dom tree to copy into a tree model.
+   * @param {Object} data - Data to attach to the root node.
+   * @param {Function} procFunc - Function to call on each dom node.
+   * @returns {TreeDomModel} An instance.
+   */
+  function init( dom, data, selector, procFunc ) {
+    _tree = new Tree();
+    _tree.init( data );
+    _populateTreeFromDom( dom, _tree.getRoot(), selector, procFunc );
+
+    return this;
+  }
+
+  /**
+   * Perform a recursive depth-first search of the DOM
+   * and call a function for each node.
+   * @param {HTMLNode} dom - A DOM element to search from.
+   * @param {TreeNode} parentNode
+   *   Node in a tree from which to attach new nodes.
+   * @param {TreeNode} selector
+   *   A CSS selector to check that DOM nodes match against,
+   *   if they match, include the node in the tree.
+   * @param {Function} callback - Function to call on each node.
+   */
+  function _populateTreeFromDom( dom, parentNode, selector, callback ) {
+    const children = dom.children;
+    // IE11 does not have children on SVG nodes.
+    if ( !children ) {
+      return;
+    }
+    let child;
+    for ( let i = 0, len = children.length; i < len; i++ ) {
+      child = children[i];
+      let childNode = parentNode;
+      // If the element matches the selector, add a tree node and run callback.
+      const matchesSelector = _getMatchesMethod( child );
+      if ( matchesSelector.bind( child )( selector ) ) {
+        childNode = parentNode.tree.add( null, parentNode );
```

Comment:
@jimmynotjim this is the one area I'm not so into with this update. As the tree nodes are created and added to the tree their data payload is empty (`null`), since we don't know at this point that they will be holding menus (FlyoutMenu instances), the data payload is then set with [`treeNode.data = menu`](https://github.com/cfpb/cfgov-refresh/pull/4316/files#diff-7c890f8c0c7dc76e2ab1fba0db2d8f64R113) in the `_addMenu` function in the MegaMenu module. Consequently, the data payload is touched twice, once here to set it to null initially and once to add the menu when a new FlyoutMenu instance is created. Before creation of the menu and creation of the new tree node happened in the same function (`_addMenu`) so the data payload could be set at the same point, so this two-step process to setting it is a consequence of separating tree node creation from FlyoutMenu creation.